worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java

```diff
@@ -63,57 +104,249 @@ public WorkerQueueConsumerImpl(TaskCallback callback, RabbitMetricsReporter metr
+                    routingKey,
+                    deliveryHeaders,
+                    taskMessage,
+                    deliveryMessageData,
+                    poisonMessageStatus == PoisonMessageStatus.POISON
+            );
+        } catch (final TransientDeliveryException e) {
+            LOG.warn("Transient error processing message id {}, disconnecting.", inboundMessageId, e);
+            offloadedPayloadsToDelete.remove(inboundMessageId);
+            //Disconnect the channel to allow for a reconnect when the HealthCheck passes.
+            disconnectCallback.run();
+        }
+    }
+
+    /**
+     * Handles the logic for injecting taskData from the store into the TaskMessage if required.
+     * Returns true if processing should continue, false if it should stop (e.g. error).
+     * If invalid, handles as poison message (publishes to retry queue) and returns false.
+     */
+    private void handleTaskDataInjection(final TaskMessage taskMessage, final long inboundMessageId,
+                                         final Optional<String> taskMessageStorageRefOpt) 
+        throws InvalidDeliveryException, TransientDeliveryException
+    {
+        final byte[] currentTaskData = taskMessage.getTaskData();
+        final boolean hasStorageRef = taskMessageStorageRefOpt.isPresent();
+        final boolean hasTaskData = currentTaskData != null;
+
+        if (hasTaskData && hasStorageRef) {
+            throw new InvalidDeliveryException(
+                    "TaskMessage contains both taskData and a storage reference. This is invalid.", inboundMessageId);
+        }
+        if (!hasTaskData && !hasStorageRef) {
+            throw new InvalidDeliveryException(
+                    "TaskMessage contains neither taskData nor a storage reference. This is invalid.", inboundMessageId);
+        }
+        if (hasStorageRef) {
+            final byte[] offloadedTaskData;
+            try {
+                offloadedTaskData = retrieveTaskDataFromStore(taskMessageStorageRefOpt.get(), inboundMessageId);
+            } catch (final ReferenceNotFoundException e) {
+                throw new InvalidDeliveryException(e.getMessage(), inboundMessageId);
+            }
+            taskMessage.setTaskData(offloadedTaskData);
+        }
+        // If hasTaskData and !hasStorageRef, nothing to do
+    }
+
+    private byte[] retrieveTaskDataFromStore(final String taskMessageStorageRef, final long inboundMessageId) 
+            throws ReferenceNotFoundException, TransientDeliveryException
+    {
+        try (final var inputStream = dataStore.retrieve(taskMessageStorageRef)) {
+            final var taskData = inputStream.readAllBytes();
+            offloadedPayloadsToDelete.put(inboundMessageId, taskMessageStorageRef);
+            return taskData;
+        } catch (final IOException | DataStoreException ex) {
+            if (ex instanceof ReferenceNotFoundException) {
+                throw (ReferenceNotFoundException)ex;
+            }
```

Comment:
Why is the `ReferenceNotFoundException` not just being caught first?

```suggestion
        } catch (final ReferenceNotFoundException ex) {
            throw ex;
        } catch (final IOException | DataStoreException ex) {
```